### ISSUE TITLE
[BUG] Linux version not detecting Streamdeck

### ISSUE THREAD

**Issue by JonnyTech** [bitfocus/companion#1279]
No device listed

Follow instructions at https://github.com/bitfocus/companion/wiki/Installation
Rules file: `/etc/udev/rules.d/50-companion.rules`

Latest version `companion-2.1.1-acb83e9-2660-linux.tar.gz` on Debian 10

```
$ lsusb |grep Elgato
Bus 002 Device 005: ID 0fd9:006c Elgato Systems GmbH 

$ dmesg
usb 2-2: New USB device found, idVendor=0fd9, idProduct=006c, bcdDevice= 2.00
usb 2-2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
usb 2-2: Product: Stream Deck XL
usb 2-2: Manufacturer: Elgato
usb 2-2: SerialNumber: <removed>
input: Elgato Stream Deck XL as /devices/pci0000:00/0000:00:14.0/usb2/2-2/2-2:1.0/0003:0FD9:006C.0007/input/input26
hid-generic 0003:0FD9:006C.0007: input,hidraw0: USB HID v1.10 Device [Elgato Stream Deck XL] on usb-0000:00:14.0-2/input0

$ /opt/companion/companion 
module.js:545
    throw err;
    ^

Error: Cannot find module 'detect-libc'
    at Function.Module._resolveFilename (module.js:543:15)
    at Function.Module._load (module.js:473:25)
    at Module.require (module.js:586:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/opt/companion/resources/node_modules/sharp/lib/platform.js:3:20)
    at Object.<anonymous> (/opt/companion/resources/node_modules/sharp/lib/platform.js:26:3)
    at Module._compile (module.js:642:30)
    at Object.Module._extensions..js (module.js:653:10)
    at Module.load (module.js:561:32)
    at tryModuleLoad (module.js:504:12)
```

![sd-nq8](https://user-images.githubusercontent.com/3584732/98862934-9548c780-245f-11eb-8f0e-4eeb64aca25b.png)

**Comment by josephdadams** [comment]
@jarodwsams

**Comment by jarodwsams** [comment]
Looks like there's a dependency missing. Try running this:
```
sudo apt-get install libgusb-dev npm nodejs git build-essential cmake libudev-dev libusb-1.0-0-dev -y
```
This is the dependency line from the Raspberry Pi instructions, but that's based on Debian so should do the trick.

**Comment by JonnyTech** [comment]
Thanks @jarodwsams, I installed the dependencies and rebooted but that made no difference. Can you suggest anything else?

**Comment by jarodwsams** [comment]
Try the UDEV from RPi. I know the UDEV in the Linux instructions _should_ be wildcard and cover everything, but just to be sure:
```
SUBSYSTEM=="input", GROUP="input", MODE="0666"
SUBSYSTEM=="usb", ATTRS{idVendor}=="0fd9", ATTRS{idProduct}=="0060", MODE:="666", GROUP="plugdev"
KERNEL=="hidraw", ATTRS{idVendor}=="0fd9", ATTRS{idProduct}=="0060", MODE:="666", GROUP="plugdev"
SUBSYSTEM=="usb", ATTRS{idVendor}=="ffff", ATTRS{idProduct}=="1f40", MODE:="666", GROUP="plugdev"
KERNEL=="hidraw", ATTRS{idVendor}=="ffff", ATTRS{idProduct}=="1f40", MODE:="666", GROUP="plugdev"
SUBSYSTEM=="usb", ATTRS{idVendor}=="0fd9", ATTRS{idProduct}=="0063", MODE:="666", GROUP="plugdev"
KERNEL=="hidraw", ATTRS{idVendor}=="0fd9", ATTRS{idProduct}=="0063", MODE:="666", GROUP="plugdev"
SUBSYSTEM=="usb", ATTRS{idVendor}=="0fd9", ATTRS{idProduct}=="006c", MODE:="666", GROUP="plugdev"
KERNEL=="hidraw", ATTRS{idVendor}=="0fd9", ATTRS{idProduct}=="006c", MODE:="666", GROUP="plugdev"
SUBSYSTEM=="usb", ATTRS{idVendor}=="0fd9", ATTRS{idProduct}=="006d", MODE:="666", GROUP="plugdev"
KERNEL=="hidraw", ATTRS{idVendor}=="0fd9", ATTRS{idProduct}=="006d", MODE:="666", GROUP="plugdev"
SUBSYSTEM=="usb", ATTRS{idVendor}=="ffff", ATTRS{idProduct}=="1f41", MODE:="666", GROUP="plugdev"
KERNEL=="hidraw", ATTRS{idVendor}=="ffff", ATTRS{idProduct}=="1f41", MODE:="666", GROUP="plugdev"
```

**Comment by jarodwsams** [comment]
Oh, and have you installed npm and node?
```
sudo npm install n -g
sudo n 8.12.0
```

**Comment by JonnyTech** [comment]
@jarodwsams, thanks, I already tried the pi rules, and node/npm were part or the original installation instructions. Booting to Windows works so not a hardware issue. Are there any logs to examine?

**Comment by elratt0r** [comment]
I investigated a bit, though I am more or less unfamiliar with nodejs and the build system.

**Problem 1:**
in _resources/node_modules/sharp/vendor/lib_ the files that should be symlinks
are text files containing the link destination, for example:
libz.so and libz.so.1 contain "libz.so.1.2.11" instead of being a symlink to libz.so.1.2.11

**Problem 2:**
some node_modules are missing in the build:
color
color-convert
color-name
color-string
detect-libc
is-arrayish
simple-swizzle

Quick and dirty hack to make it work (assuming you're in the companion root folder):
```
cd resources/node_modules/sharp/vendor/lib/
for i in `find . -type f -regex './lib.*\.so[.]?[0-9]*$'`; do LINK=`cat $i`; echo "ln -s $LINK $i"; ln -fs $LINK $i; done
cd ../../../../../

cd resources/node_modules/
git clone https://github.com/Qix-/color color
git clone https://github.com/Qix-/color-convert color-convert
git clone https://github.com/colorjs/color-name color-name
git clone https://github.com/Qix-/color-string color-string
git clone https://github.com/lovell/detect-libc detect-libc
git clone https://github.com/Qix-/node-is-arrayish is-arrayish
git clone https://github.com/Qix-/node-simple-swizzle simple-swizzle
cd ../../
```

then you should be able to run companion without any errors.

Hope that helps a bit.

**Comment by Julusian** [comment]
It looks like those are all the sharp dependencies that are bundled inside the asar, but when loading sharp it is looking in the wrong place so cant find then.

Do either of 2.0.0 or 2.1.0 work for you?
This looks to be caused by a change of mine (https://github.com/bitfocus/companion/commit/01facfa63bae46a33c7544afe40d58848dae47e4), which claims to fix some sharp issues...

**Comment by JonnyTech** [comment]
Thanks guys, I shall test when I am next with the hardware - hopefully next week. I guess that the  `detect-libc` node module would restore hardware detection.

**Comment by elratt0r** [comment]
> Do either of 2.0.0 or 2.1.0 work for you?

Could not find any 2.0.0 release or candidate to download.
Tried (from https://bitfocus.io/companion/download/builds/):
companion-2.0.1-96d0a44-2484-linux.tar.gz
companion-2.0.1-24822ac-2486-linux.tar.gz
companion-2.0.1-1c0e74d-2491-linux.tar.gz
companion-2.1.0-47823d0-2493-linux.tar.gz

2484 crashes with "Cannot find module '@sentry/electron'"
2486, 2491 and 2493 as described above ("module.js:545; throw err;" and "detect-libc missing")

**Comment by stevespaw** [comment]
I am a TOTAL linux noob, but I have the same issue. I have put in the rules file. Newest build today. I get this when opening:
xxx@zzz-source:~/Downloads/companion-2.2.0$ ./companion
(node:3045) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(node:3045) Warning: require() of ES modules is not supported.
require() of /home/ndi/Downloads/companion-2.2.0/resources/app.asar/lib/module/homeassistant-server/node_modules/home-assistant-js-websocket/dist/haws.umd.js from /home/ndi/Downloads/companion-2.2.0/resources/app.asar/lib/module/homeassistant-server/dist/index.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
Instead rename haws.umd.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /home/ndi/Downloads/companion-2.2.0/resources/app.asar/lib/module/homeassistant-server/node_modules/home-assistant-js-websocket/package.json.

**Comment by Julusian** [comment]
@elratt0r, this is resolved in the 2.2.0 builds. Quite a few dependencies were changed as well as how some of the bundling is performed which could be helping here. I don't know if it is worth looking into any further in 2.1.1

@stevespaw, those are all unrelated errors that you can safely ignore

**Comment by Fechtenburg** [comment]
Some time ago, I gave it a shot, but didn't succeed to have my Stream Deck connected.

The other day, I gave it a try again, and in no time I was up and running.

Then I copied the image to another Raspberry PI, and the Stream Deck was not found.

I tried different things, connected different Stream Decks and so on.

The issue in my case was the USB cable from IKEA. Just by replacing the cable, the Stream Deck was found instantly.

If you have problems connecting to the Stream Deck, changing USB cable might help.
